cura/Machines/QualityManager.py

```diff
@@ -0,0 +1,461 @@
+# Copyright (c) 2018 Ultimaker B.V.
+# Cura is released under the terms of the LGPLv3 or higher.
+
+from typing import Optional, List
+
+from PyQt5.QtCore import QObject, QTimer, pyqtSignal, pyqtSlot
+
+from UM.Application import Application
+from UM.Logger import Logger
+from UM.Util import parseBool
+
+from cura.Machines.ContainerNode import ContainerNode
+
+
+#
+# Quality lookup tree structure:
+#
+#      <machine_definition_id>------|
+#                |                  |
+#          <variant_name>        <root_material_id>
+#                |
+#        <root_material_id>
+#                |
+#          <quality_type>
+#                |
+#            <quality_name>
+#       + <quality_changes_name>
+#
+
+#
+# A QualityGroup represents a group of containers that must be applied to each ContainerStack when it's used.
+# Some concrete examples are Quality and QualityChanges: when we select quality type "normal", this quality type
+# must be applied to all stacks in a machine, although each stack can have different containers. Use an Ultimaker 3
+# as an example, suppose we choose quality type "normal", the actual InstanceContainers on each stack may look
+# as below:
+#                       GlobalStack         ExtruderStack 1         ExtruderStack 2
+# quality container:    um3_global_normal   um3_aa04_pla_normal     um3_aa04_abs_normal
+#
+# This QualityGroup is mainly used in quality and quality_changes to group the containers that can be applied to
+# a machine, so when a quality/custom quality is selected, the container can be directly applied to each stack instead
+# of looking them up again.
+#
+class QualityGroup(QObject):
```

Comment:
Code style: Class name doesn't match file name.

Reply:
Split into separate files in 97059daf6